File: cs/TestECDH.cs
using System;

/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

/* test driver and function exerciser for ECDH/ECIES/ECDSA API Functions */

public class TestECDH
{
	public static void printBinary(sbyte[] array)
	{
		int i;
		for (i = 0;i < array.Length;i++)
		{
			Console.Write("{0:x2}", array[i]);
		}
		Console.WriteLine();
	}

	public static void Main(string[] args)
	{
		int i , j = 0, res ;
		int result;
		string pp = "M0ng00se";

		int EGS = ECDH.EGS;
		int EFS = ECDH.EFS;
		int EAS = AES.KS;

		sbyte[] S1 = new sbyte[EGS];
		sbyte[] W0 = new sbyte[2 * EFS + 1];
		sbyte[] W1 = new sbyte[2 * EFS + 1];
		sbyte[] Z0 = new sbyte[EFS];
		sbyte[] Z1 = new sbyte[EFS];
		sbyte[] RAW = new sbyte[100];
		sbyte[] SALT = new sbyte[8];
		sbyte[] P1 = new sbyte[3];
		sbyte[] P2 = new sbyte[4];
		sbyte[] V = new sbyte[2 * EFS + 1];
		sbyte[] M = new sbyte[17];
		sbyte[] T = new sbyte[12];
		sbyte[] CS = new sbyte[EGS];
		sbyte[] DS = new sbyte[EGS];

		RAND rng = new RAND();

		rng.clean();
		for (i = 0;i < 100;i++)
		{
			RAW[i] = (sbyte)(i);
		}

		rng.seed(100,RAW);

//for (j=0;j<100;j++)
//{

		for (i = 0;i < 8;i++)
		{
			SALT[i] = (sbyte)(i + 1); // set Salt
		}

		Console.WriteLine("Alice's Passphrase= " + pp);
		sbyte[] PW = pp.GetBytes();

/* private key S0 of size EGS bytes derived from Password and Salt */

		sbyte[] S0 = ECDH.PBKDF2(PW,SALT,1000,EGS);

		Console.Write("Alice's private key= 0x");
		printBinary(S0);

/* Generate Key pair S/W */
		ECDH.KEY_PAIR_GENERATE(null,S0,W0);

		Console.Write("Alice's public key= 0x");
		printBinary(W0);

		res = ECDH.PUBLIC_KEY_VALIDATE(true,W0);
		if (res != 0)
		{
			Console.WriteLine("ECP Public Key is invalid!\n");
			return;
		}
/* Random private key for other party */
		ECDH.KEY_PAIR_GENERATE(rng,S1,W1);

		Console.Write("Servers private key= 0x");
		printBinary(S1);

		Console.Write("Servers public key= 0x");
		printBinary(W1);


		res = ECDH.PUBLIC_KEY_VALIDATE(true,W1);
		if (res != 0)
		{
			Console.Write("ECP Public Key is invalid!\n");
			return;
		}

/* Calculate common key using DH - IEEE 1363 method */

		ECDH.ECPSVDP_DH(S0,W1,Z0);
		ECDH.ECPSVDP_DH(S1,W0,Z1);

		bool same = true;
		for (i = 0;i < EFS;i++)
		{
			if (Z0[i] != Z1[i])
			{
				same = false;
			}
		}

		if (!same)
		{
			Console.WriteLine("*** ECPSVDP-DH Failed");
			return;
		}

		sbyte[] KEY = ECDH.KDF1(Z0,EAS);

		Console.Write("Alice's DH Key=  0x");
		printBinary(KEY);
		Console.Write("Servers DH Key=  0x");
		printBinary(KEY);

		Console.WriteLine("Testing ECIES");

		P1[0] = 0x0;
		P1[1] = 0x1;
		P1[2] = 0x2;
		P2[0] = 0x0;
		P2[1] = 0x1;
		P2[2] = 0x2;
		P2[3] = 0x3;

		for (i = 0;i <= 16;i++)
		{
			M[i] = (sbyte)i;
		}

		sbyte[] C = ECDH.ECIES_ENCRYPT(P1,P2,rng,W1,M,V,T);

		Console.WriteLine("Ciphertext= ");
		Console.Write("V= 0x");
		printBinary(V);
		Console.Write("C= 0x");
		printBinary(C);
		Console.Write("T= 0x");
		printBinary(T);


		M = ECDH.ECIES_DECRYPT(P1,P2,V,C,T,S1);
		if (M.Length == 0)
		{
			Console.WriteLine("*** ECIES Decryption Failed\n");
			return;
		}
		else
		{
			Console.WriteLine("Decryption succeeded");
		}

		Console.Write("Message is 0x");
		printBinary(M);

		Console.WriteLine("Testing ECDSA");

		if (ECDH.ECPSP_DSA(rng,S0,M,CS,DS) != 0)
		{
			Console.WriteLine("***ECDSA Signature Failed");
			return;
		}
		Console.WriteLine("Signature= ");
		Console.Write("C= 0x");
		printBinary(CS);
		Console.Write("D= 0x");
		printBinary(DS);

		if (ECDH.ECPVP_DSA(W0,M,CS,DS) != 0)
		{
			Console.WriteLine("***ECDSA Verification Failed");
			return;
		}
		else
		{
			Console.WriteLine("ECDSA Signature/Verification succeeded " + j);
		}
//}
//System.out.println("Test Completed Successfully");
	}
}

File: swift/TestECDH.swift
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/
//
//  TestECDH.swift
//  
//
//  Created by Michael Scott on 02/07/2015.
//  Copyright (c) 2015 Michael Scott. All rights reserved.
//

import Foundation
import clint // comment out for Xcode

public func TestECDH()
{
    let pp=String("M0ng00se");

    let EGS=ECDH.EGS
    let EFS=ECDH.EFS
    let EAS=AES.KS

    var S1=[UInt8](count:EGS,repeatedValue:0)
    var W0=[UInt8](count:2*EFS+1,repeatedValue:0)
    var W1=[UInt8](count:2*EFS+1,repeatedValue:0)
    var Z0=[UInt8](count:EFS,repeatedValue:0)
    var Z1=[UInt8](count:EFS,repeatedValue:0)
    var RAW=[UInt8](count:100,repeatedValue:0)
    var SALT=[UInt8](count:8,repeatedValue:0)
    var P1=[UInt8](count:3,repeatedValue:0)
    var P2=[UInt8](count:4,repeatedValue:0)
    var V=[UInt8](count:2*EFS+1,repeatedValue:0)
    var M=[UInt8](count:17,repeatedValue:0)
    var T=[UInt8](count:12,repeatedValue:0)
    var CS=[UInt8](count:EGS,repeatedValue:0)
    var DS=[UInt8](count:EGS,repeatedValue:0)

    let rng=RAND()

    rng.clean();
    for var i=0;i<100;i++ {RAW[i]=UInt8(i&0xff)}

    rng.seed(100,RAW)


    for var i=0;i<8;i++ {SALT[i]=UInt8(i+1)}  // set Salt

    print("Alice's Passphrase= "+pp)
    let PW=[UInt8](pp.utf8)

    /* private key S0 of size EGS bytes derived from Password and Salt */

    var S0=ECDH.PBKDF2(PW,SALT,1000,EGS)
    print("Alice's private key= 0x",terminator: ""); ECDH.printBinary(S0)

    /* Generate Key pair S/W */
    ECDH.KEY_PAIR_GENERATE(nil,&S0,&W0);

    print("Alice's public key= 0x",terminator: ""); ECDH.printBinary(W0)

    var res=ECDH.PUBLIC_KEY_VALIDATE(true,W0);

    if res != 0
    {
        print("ECP Public Key is invalid!");
        return;
    }

    /* Random private key for other party */
    ECDH.KEY_PAIR_GENERATE(rng,&S1,&W1)

    print("Servers private key= 0x",terminator: ""); ECDH.printBinary(S1)

    print("Servers public key= 0x",terminator: ""); ECDH.printBinary(W1);

    res=ECDH.PUBLIC_KEY_VALIDATE(true,W1)
    if res != 0
    {
        print("ECP Public Key is invalid!")
        return
    }

    /* Calculate common key using DH - IEEE 1363 method */

    ECDH.ECPSVDP_DH(S0,W1,&Z0)
    ECDH.ECPSVDP_DH(S1,W0,&Z1)

    var same=true
    for var i=0;i<EFS;i++
    {
        if Z0[i] != Z1[i] {same=false}
    }

    if (!same)
    {
        print("*** ECPSVDP-DH Failed")
        return
    }

    let KEY=ECDH.KDF1(Z0,EAS)

    print("Alice's DH Key=  0x",terminator: ""); ECDH.printBinary(KEY)
    print("Servers DH Key=  0x",terminator: ""); ECDH.printBinary(KEY)

    print("Testing ECIES")

    P1[0]=0x0; P1[1]=0x1; P1[2]=0x2
    P2[0]=0x0; P2[1]=0x1; P2[2]=0x2; P2[3]=0x3

    for var i=0;i<=16;i++ {M[i]=UInt8(i&0xff)}

    let C=ECDH.ECIES_ENCRYPT(P1,P2,rng,W1,M,&V,&T)

    print("Ciphertext= ")
    print("V= 0x",terminator: ""); ECDH.printBinary(V)
    print("C= 0x",terminator: ""); ECDH.printBinary(C)
    print("T= 0x",terminator: ""); ECDH.printBinary(T)

    M=ECDH.ECIES_DECRYPT(P1,P2,V,C,T,S1)
    if M.count==0
    {
        print("*** ECIES Decryption Failed\n")
        return
    }
    else {print("Decryption succeeded")}

    print("Message is 0x"); ECDH.printBinary(M)

    print("Testing ECDSA")

    if ECDH.ECPSP_DSA(rng,S0,M,&CS,&DS) != 0
    {
        print("***ECDSA Signature Failed")
        return
    }
    print("Signature= ")
    print("C= 0x",terminator: ""); ECDH.printBinary(CS)
    print("D= 0x",terminator: ""); ECDH.printBinary(DS)

    if ECDH.ECPVP_DSA(W0,M,CS,DS) != 0
    {
        print("***ECDSA Verification Failed")
        return
    }
    else {print("ECDSA Signature/Verification succeeded ")}

}

TestECDH()  // comment out for Xcode
